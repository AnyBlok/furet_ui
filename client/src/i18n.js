import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

export const i18nConf = {
  locale: 'en',
  fallbackLocale: 'en',
  messages: {
    en: {
    },
  },
};
export const i18n = new VueI18n(i18nConf);
