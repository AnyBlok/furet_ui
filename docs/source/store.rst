.. This file is a part of the FuretUI project
..
..    Copyright (C) 2014 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>
..
.. This Source Code Form is subject to the terms of the Mozilla Public License,
.. v. 2.0. If a copy of the MPL was not distributed with this file,You can
.. obtain one at http://mozilla.org/MPL/2.0/.

.. contents::

Reducer
=======

client
------

Add any information about a custom client::

    store.state.client: {
        ``custom view name``: {
            ``params of the custum view``,
        },
        ...
    }

UPDATE_CLIENT
~~~~~~~~~~~~~

Add params for a custom view::

    {
        type: 'UPDATE_VIEW_CLIENT',
        viewName: ``custom view name``,
        ``params ...``,
    }

CLEAR_CLIENT
~~~~~~~~~~~~

Clear all the params::

    {
        type: 'CLEAR_CLIENT'
    }


leftmenu, rightmenu
-------------------

It is the same for the both. The define dialog menu::

    store.state.[ leftmenu | rightmenu ] = {
        value: {
            label: ``label of the selected menu``,
            image: {
                type: [ font-icon ],
                value: ``value in function of type to display the image``,
            },
        },
        values: [
            {
                label: ``label of the selected menu``,
                description: ``Description of the menu``,
                image: {
                    type: [ font-icon ],
                    value: ``value in function of type to display the image``,
                },
                type: [ SPACE | CLIENT ]
                id: ``space id or custom view name in function of type value``
            },
        ],
    }

UPDATE_LEFT_MENU, UPDATE_RIGHT_MENU
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Update the menu definition or selected menu::

    {
        type: [ UPDATE_LEFT_MENU | UPDATE_RIGHT_MENU ],
        value: {
            label: ``label of the selected menu``,
            image: {
                type: [ font-icon ],
                value: ``value in function of type to display the image``,
            },
        },
        values: [
            {
                label: ``label of the selected menu``,
                description: ``Description of the menu``,
                image: {
                    type: [ font-icon ],
                    value: ``value in function of type to display the image``,
                },
                type: [ SPACE | CLIENT ]
                id: ``space id or custom view name in function of type value``
            },
        ],
    }

.. note::

    value and values is not required



CLEAR_LEFT_MENU, CLEAR_RIGHT_MENU
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Clear the menu::

    {
        type: [ CLEAR_LEFT_MENU | CLEAR_RIGHT_MENU ]
    }


global
------

Define the context of the application::

    store.state.global = {
        title: ``title of the application``,
        modal_custom_view: ``the current modal custom view``,
        breadscrumbs: [
            {
                path: ``the path of the previous action``,
                label: ``label display in the breadcrumb``,
                changes: ``the changes did during the action``,
                position: ``to order the Array``,
            },
            ...
        ],
    }

UPDATE_GLOBAL
~~~~~~~~~~~~~

Update the global values::

    {
        type: 'UPDATE_GLOBAL', 
        ``params``,
    }

.. warning:: 

    Used for title or modal_custom_view.


CLEAR_GLOBAL
~~~~~~~~~~~~

Clear all the parameter in global::

    {
        type: 'CLEAR_GLOBAL',
    }


ADD_IN_BREADSCRUMB
~~~~~~~~~~~~~~~~~~

Add new entry at the end of the breadcrumb::

    {
        type: 'ADD_IN_BREADSCRUMB',
        path: ``the path to save``,
        label: ``the label to display``,
        changes: ``the state data changes to save``,
    }

REMOVE_FROM_BREADSCRUMB
~~~~~~~~~~~~~~~~~~~~~~~

Remove the breadcrumb entries if the position is higther than the wanted position::

    {
        type: 'REMOVE_FROM_BREADSCRUMB',
        position: ``the position to delete``
    }

CLEAR_BREADSCRUMB
~~~~~~~~~~~~~~~~~

Clear all the breadcrumb entries::

    {
        type: 'CLEAR_BREADSCRUMB',
    }

data
----

Save all the data for furet_ui::

    store.state.data = {
        actions: {
            ``action id``: {
                label: ``Label of the action``,
                views: [
                    {
                        viewId: ``view id``,
                        type: ``type of view``,
                    },
                    ...
                ],
            },
            ...
        },
        views: {
            ``view's id``: {
                model: ``model's name``,
                ``view params which depend of the type of the view``
            },
        },
        data: {
            ``Model name``: {
                ``data id``: {
                    ``data``,
                    ...
                },
                ...
            },
            ...
        },
        changes: {
            new: {
                ``Model name``: {
                    ``data id``: [ 'DELETED' || { ``data`` } ],
                    ...
                },
                ...
            },
            ``Model name``: {
                ``data id``: [ 'DELETED' || { ``data`` } ],
                ...
            },
            ...
        },
        spaces: {
            left_menu: [
                ``menu params``,
                ...
            ],
            right_menu: [
                ``menu params``,
                ...
            ],
        }
    }

the left and right menu definition is the same::

    [
        {
            label: ``displayed label``,
            image: {
                type: [ 'font-icon' ],
                value: ``value of the icon``,             
            actionId: ``action id``,
            id: ``id of the menu``,                                                      
            submenus: [
                ``Same menu definition, to use the submenus don't put actionId``,
                ...
            ],                                                 
        },                                                                  


UPDATE_ACTION
~~~~~~~~~~~~~

Add action definition::

    {
        type: 'UPDATE_ACTION_MANAGER_ADD_ACTION_DATA',
        actionId: ``action id``
        label: ``Label of the action``,
        views: [
            {
                viewId: ``view id``,
                type: ``type of view``,
            },
            ...
        ],
    }


UPDATE_VIEW
~~~~~~~~~~~

Update the data of one view::

    {
        type: 'UPDATE_VIEW',
        viewId: ``view id``,
        model: ``model name``,
        ``params of the view``,
    }

UPDATE_DATA
~~~~~~~~~~~

Update the data::

    {
        type: 'UPDATE_DATA',
        model: ``Model name``,
        data: {
            ``data id``: {
                ``params``,
            },
            ...
        },
    }

DELETE_DATA
~~~~~~~~~~~

Delete some data::

    {
        type: 'DELETE_DATA',
        data: {
            ``Model name``: [
                ``data id``,
            ],
        },
    }

UPDATE_CHANGE
~~~~~~~~~~~~~

Modify current change::

    {
        type: 'UPDATE_CHANGE',
        model: ``Model name``,
        dataId: ``data id``,
        fieldname: ``name of the field``,
        value: ``new value to save``,
    }

REPLACE_CHANGE
~~~~~~~~~~~~~~

replace the change by another::

    {
        type: 'REPLACE_CHANGE',
        changes: ``object``,
    }

CLEAR_CHANGE
~~~~~~~~~~~~

replace the change by another::

    {
        type: 'CLEAR_CHANGE',
    }

CREATE_CHANGE_X2M
~~~~~~~~~~~~~~~~~

Add new change in the new entry, use by One2Many and Many2Many::

    {
        type: 'CREATE_CHANGE_X2M',
        model: ``model name``,
        dataId: ``id of the data to create``
    }

UPDATE_CHANGE_X2M
~~~~~~~~~~~~~~~~~

Modify current change in new if exist or in the normal place::

    {
        type: 'UPDATE_CHANGE_X2M',
        model: ``Model name``,
        dataId: ``data id``,
        fieldname: ``name of the field``,
        value: ``new value to save``,
    }

UPDATE_CHANGE_X2M_DELETE
~~~~~~~~~~~~~~~~~~~~~~~~

Remove the change in the new entry if exist else add the 'DELETED' in the nomal place for the model and the dataIds::

    {
        type: 'UPDATE_CHANGE_X2M_DELETE',
        model: ``Model name``,
        dataIds: [
            ``data id``,
            ...
        ],
    }

UPDATE_SPACE
~~~~~~~~~~~~

Update the space data::

    {
        type: 'UPDATE_SPACE',
        spaceId: ``
        left_menu: [
            ``menu params``,
            ...
        ],
        right_menu: [
            ``menu params``,
            ...
        ],
    }

CLEAR_DATA
~~~~~~~~~~

Clear all the data::

    {
        type: 'CLEAR_DATA'
    }

View definition for store
-------------------------

.. note::

    the params are different for each view type

List
~~~~

::

    label: ``label of the view``,
    model: ``model name``,
    creatable: ``true display the create button``,
    deletable: ``true display the delete button``,
    selectable: ``true display the check box``,
    onSelect: ``view id to use if the line is clicked``,
    headers: [
        {
            name: ``column name``,
            label: ``Label of the column``,
            component: ``the vue component name``
            sortable: ``true: the column is sortable``
            numeric: ``true the value is a numeric``
            invisible: ``condition to determinate if the field is displayed or not``
        },
        ...
    ],
    search: [
        {
            key: ``key used by the server to filter the data``,
            label: ``Label display of the key``,
            type: ``what ever if you need to split the filter``,
            operator: ``what ever if you want define this``,
            value: ``the value``,
        },
        ...
    ],
    buttons: [
        {
            label: ``Label of the button``,
            buttonId: ``id of the button``,
        },
        ...
    ],
    onSelect_buttons: [
        {
            label: ``Label of the button``,
            buttonId: ``id of the button``,
        },
        ...
    ],
    fields: [
        ``column name``,
        ...
    ],

Thumbnail
~~~~~~~~~

::

    label: ``label of the view``,
    model: ``model name``,
    creatable: ``true display the create button``,
    deletable: ``true display the delete button``,
    onSelect: ``view id to use if the line is clicked``,
    search: [
        {
            key: ``key used by the server to filter the data``,
            label: ``Label display of the key``,
            type: ``what ever if you need to split the filter``,
            operator: ``what ever if you want define this``,
            value: ``the value``,
        },
        ...
    ],
    template: ``template definition``,
    buttons: [
        {
            label: ``Label of the button``,
            buttonId: ``id of the button``,
        },
        ...
    ],
    fields: [
        ``column name``,
        ...
    ],


.. note::

    For the template see the template page :ref:`template`


Form
~~~~

::

    label: ``label of the view``,
    model: ``model name``,
    creatable: ``true display the create button``,
    deletable: ``true display the delete button``,
    editable: ``true display the edit button``,
    onClose: ``view id to use if the close button is clicked``,
    template: ``template definition``,
    buttons: [
        {
            label: ``Label of the button``,
            buttonId: ``id of the button``,
        },
        ...
    ],
    fields: [
        ``column name``,
        ...
    ],


.. note::

    For the template see the template page :ref:`template`
