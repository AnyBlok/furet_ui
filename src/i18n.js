/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';
import VueI18n from 'vue-i18n';
Vue.use(VueI18n);


export const i18nConf = {
    locale: 'en',
    fallbackLocale: 'en',
    messages: {
        en: {
            menus: {
                close: 'Close',
                search: 'Search ...',
            },
            views: {
                unknown: {
                    title: 'The wanted view "{name}" is unknown',
                    message: "Report this message to the administrator",
                },
                common: {
                    create: 'Create',
                    save: 'Save',
                    edit: 'Edit',
                    cancel: 'Cancel',
                    delete: 'Delete',
                    close: 'Close',
                    actions: 'Actions',
                    more: 'More',
                },
                clients: {
                    login: {
                        button: 'Log in',
                    },
                },
            },
            fields: {
                common: {
                    required: 'this field is required',
                },
                json: {
                    invalid: 'Invalid JSON format',
                },
                file: {
                    upload: 'Upload',
                    download: 'Download',
                    delete: 'Delete',
                },
            },
        },
    },
};
export const i18n = new VueI18n(i18nConf);
