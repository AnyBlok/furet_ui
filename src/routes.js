/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';
import VueRouter from 'vue-router';
Vue.use(VueRouter);
import {json_post_dispatch_all} from './server-call';


const CustomeView = {
    template: '<furet-ui-custom-view v-bind:viewName="viewName" />',
    props: ['viewName'],
    beforeRouteEnter: (to, from, next) => {
        json_post_dispatch_all('/custom/view/' + to.params.viewName, {});
        next();
    },
    beforeRouteUpdate: (to, from, next) => {
        if (to.params.viewName != from.params.viewName) {
            json_post_dispatch_all('/custom/view/' + to.params.viewName, {});
        }
        next();
    },
}

const Action = {
    template: '<router-view></router-view>',
    beforeRouteEnter: (to, from, next) => {
        json_post_dispatch_all('/action/' + to.params.actionId, to.params);
        next();
    },
    beforeRouteUpdate: (to, from, next) => {
        if (to.params.actionId != from.params.actionId)
            json_post_dispatch_all('/action/' + to.params.actionId, to.params);
        next();
    },
};

const View = {
    template: `<furet-ui-view
                    v-bind:spaceId="spaceId"
                    v-bind:menuId="menuId"
                    v-bind:actionId="actionId"
                    v-bind:viewId="viewId"
                    v-bind:dataId="dataId"
                    v-bind:mode="mode"
                />`,
    props: ['spaceId', 'menuId', 'actionId', 'viewId', 'dataId', 'mode'],
    beforeRouteEnter: (to, from, next) => {
        json_post_dispatch_all('/view/' + to.params.viewId, to.params);
        next();
    },
    beforeRouteUpdate: (to, from, next) => {
        if (to.params.viewId != from.params.viewId)
            json_post_dispatch_all('/view/' + to.params.viewId, to.params);
        next();
    },
};


export const router = new VueRouter({
    routes: [
        {
            path: '/custom/view/:viewName',
            name: 'custom_view',
            props: true,
            component: CustomeView,
        },
        {
            path: '/space/:spaceId',
            name: 'space',
            props: true,
            component: {
                template: `<furet-ui-space
                                v-bind:spaceId="spaceId"
                                v-bind:menuId="menuId"
                                v-bind:actionId="actionId"
                                v-bind:viewId="viewId"
                            />`,
                props: ['spaceId', 'menuId', 'actionId', 'viewId'],
                beforeRouteEnter: (to, from, next) => {
                    json_post_dispatch_all('/space/' + to.params.spaceId, to.params);
                    next();
                },
                beforeRouteUpdate: (to, from, next) => {
                    if (to.params.spaceId != from.params.spaceId) json_post_dispatch_all('/space/' + to.params.spaceId, to.params);
                    next();
                },
            },
            children: [
                {
                    path: 'menu/:menuId',
                    name: 'space_menu',
                    component: {
                        template: '<router-view></router-view>'
                    },
                    children: [
                        {
                            path: 'action/:actionId',
                            name: 'space_menu_action',
                            component: Action,
                            children: [
                                {
                                    path: 'view/:viewId',
                                    name: 'space_menu_action_view',
                                    props: true,
                                    component: View,
                                    children: [
                                        {
                                            path: 'data/:dataId/mode/:mode',
                                            name: 'space_menu_action_view_dataId',
                                            component: {template: '<div />'},  // never called
                                        },
                                    ],
                                },
                                {
                                    path: 'custom/view/:viewName',
                                    name: 'space_menu_action_custom_view',
                                    props: true,
                                    component: CustomeView,
                                },
                            ],
                        },
                    ],
                },
                {
                    path: 'action/:actionId',
                    name: 'space_action',
                    component: Action,
                    children: [
                        {
                            path: 'view/:viewId',
                            name: 'space_action_view',
                            props: true,
                            component: View,
                            children: [
                                {
                                    path: 'data/:dataId/mode/:mode',
                                    name: 'space_action_view_dataId',
                                    component: {template: '<div />'},  // never called
                                },
                            ],
                        },
                        {
                            path: 'custom/view/:viewName',
                            name: 'space_action_custom_view',
                            props: true,
                            component: CustomeView,
                        },
                    ],
                },
            ]
        },
    ],
});

export default router;
