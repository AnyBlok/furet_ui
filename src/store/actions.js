/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
export const UNITEST_CLEAR = ({ commit }) => {
    commit('CLEAR_CLIENT');
    commit('CLEAR_DATA');
    commit('CLEAR_GLOBAL');
    commit('CLEAR_LEFT_MENU');
    commit('CLEAR_RIGHT_MENU');
}
