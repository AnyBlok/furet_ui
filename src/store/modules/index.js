/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import global from './global';
import client from './client';
import rightmenu from './right-menu';
import leftmenu from './left-menu';
import data from './data';

export default {
    global,
    client,
    rightmenu,
    leftmenu,
    data,
}
