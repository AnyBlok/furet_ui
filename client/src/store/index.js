import Vue from 'vue';
import Vuex from 'vuex';
import * as actions from './actions';
import modules from './modules';

Vue.use(Vuex);
const debug = process.env.NODE_ENV !== 'production';

export const storeDef = {
  actions,
  modules,
  strict: debug,
};

export const store = new Vuex.Store(storeDef);
export default store;
