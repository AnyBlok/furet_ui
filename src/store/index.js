/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import _ from 'underscore';
import Vue from 'vue';
import Vuex from 'vuex';
import * as actions from './actions';
import * as getters from './getters';
import modules from './modules';
import {i18n} from '../i18n';
import {router} from '../routes';
import moment from 'moment';

Vue.use(Vuex)
const debug = process.env.NODE_ENV !== 'production'

export const storeDef = {
    actions,
    getters,
    modules,
    strict: debug,
};

export const store = new Vuex.Store(storeDef);

export const dispatchAll = (datas) => {
    _.each(datas, data => {
        if (data.type) {
            switch (data.type) {
                case 'UPDATE_LOCALES':
                    _.each(data.locales, locale => {
                        i18n.setLocaleMessage(locale.locale, locale.messages)
                    });
                    break
                case 'SET_LOCALE':
                    i18n.locale = data.locale;
                    moment.locale(data.locale);
                    break
                case 'UPDATE_ROUTE':
                    if (data.name) router.push({name: data.name, params: data.params});
                    else if (data.path) router.push({path: data.path});
                    break;
                default:
                    const type = data.type;
                    delete data.type;
                    const process = data.process || 'commit';
                    store[process](type, data);
            }
        }
    });
}

export default store;
