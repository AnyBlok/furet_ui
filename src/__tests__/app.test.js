/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';
import {store} from '../store';
import {router} from '../routes';
import {i18n} from '../i18n';
import App from '../app';

describe('App component', () => {
    beforeEach(() => {
        store.dispatch('UNITEST_CLEAR');
    });
    it('Render App with default value from store', () => {
        const renderer = require('vue-server-renderer').createRenderer();
        const vm = new Vue({
            el: document.createElement('div'),
            store,
            router,
            i18n,
            render: h => h(App),
        });
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
    });
    it('Render App with value from store', () => {
        const renderer = require('vue-server-renderer').createRenderer();
        const vm = new Vue({
            el: document.createElement('div'),
            store,
            router,
            i18n,
            render: h => h(App),
        });
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
        store.commit('UPDATE_GLOBAL', {title: 'Test'})
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
    });
});
