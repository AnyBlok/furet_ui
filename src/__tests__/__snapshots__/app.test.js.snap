// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Render App with default value from store 1`] = `
<div data-server-rendered="true">
    <section class="hero is-primary">
        <nav class="nav">
            <div class="nav-left">
                <!---->
            </div>
            <div class="nav-center">

            </div>
            <div class="nav-right">
                <!---->
            </div>
        </nav>
    </section>
    <!---->
</div>
`;

exports[`App component Render App with value from store 1`] = `
<div data-server-rendered="true">
    <section class="hero is-primary">
        <nav class="nav">
            <div class="nav-left">
                <!---->
            </div>
            <div class="nav-center">

            </div>
            <div class="nav-right">
                <!---->
            </div>
        </nav>
    </section>
    <!---->
</div>
`;

exports[`App component Render App with value from store 2`] = `
<div data-server-rendered="true">
    <section class="hero is-primary">
        <nav class="nav">
            <div class="nav-left">
                <!---->
            </div>
            <div class="nav-center">
                Test
            </div>
            <div class="nav-right">
                <!---->
            </div>
        </nav>
    </section>
    <!---->
</div>
`;
